### PR TITLE
feat (provider/xai): Initial xAI provider.

### PR DESCRIPTION
**Pull request by shaper**

### CODE REVIEW

File: content/providers/01-ai-sdk-providers/22-grok.mdx [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+title: Grok
+description: Learn how to use Grok.
+---
```
Comment (lgrammel) [inline content/providers/01-ai-sdk-providers/22-grok.mdx:4]: - need to remove the other x.ai provider page
- might be useful to say X.ai Grok in the title & below to make it more searchable and to distinguish from Groq

File: content/docs/02-foundations/02-providers-and-models.mdx [hunk 2]
```diff
@@ -74,6 +75,7 @@ Here are the capabilities of popular models:
 | [Google Generative AI](/providers/ai-sdk-providers/google-generative-ai) | `gemini-1.5-pro`             | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> |
 | [Google Vertex](/providers/ai-sdk-providers/google-vertex)               | `gemini-1.5-flash`           | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> |
 | [Google Vertex](/providers/ai-sdk-providers/google-vertex)               | `gemini-1.5-pro`             | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> |
+| [Grok](/providers/ai-sdk-providers/grok)                                 | `grok-beta`                  | <Cross size={18} /> | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> |
```
Comment (lgrammel) [inline content/docs/02-foundations/02-providers-and-models.mdx:78]: Prefer X.ai Grok to distinguish from Groq

File: content/docs/02-foundations/02-providers-and-models.mdx [hunk 3]
```diff
@@ -31,6 +31,7 @@ The AI SDK comes with several providers that you can use to interact with differ
 - [Google Generative AI Provider](/providers/ai-sdk-providers/google-generative-ai) (`@ai-sdk/google`)
 - [Google Vertex Provider](/providers/ai-sdk-providers/google-vertex) (`@ai-sdk/google-vertex`)
 - [Mistral Provider](/providers/ai-sdk-providers/mistral) (`@ai-sdk/mistral`)
+- [Grok](/providers/ai-sdk-providers/grok) (`@ai-sdk/grok`)
```
Comment (lgrammel) [inline content/docs/02-foundations/02-providers-and-models.mdx:34]: Is grok just their model name, like claude? if so we should name the provider `xai` - they probably will have more models in the future

File: examples/ai-core/.env.example [hunk 4]
```diff
@@ -9,8 +9,9 @@ FIREWORKS_API_KEY=""
 GOOGLE_GENERATIVE_AI_API_KEY=""
 GOOGLE_VERTEX_LOCATION=""
 GOOGLE_VERTEX_PROJECT=""
+GROK_API_KEY=""
```
Comment (lgrammel) [inline examples/ai-core/.env.example:12]: use XAI api key

**Review by lgrammel — CHANGES_REQUESTED** [review]
Needs renaming

File: content/providers/01-ai-sdk-providers/22-grok.mdx (hunk 1 above)
Comment (shaper) [reply content/providers/01-ai-sdk-providers/22-grok.mdx:4]: They seem to brand themselves as `xAI` at https://x.ai/about and https://x.ai/ so I'll use that. Will rm the other and update as you note here.

File: examples/ai-core/.env.example (hunk 4 above)
Comment (shaper) [reply examples/ai-core/.env.example:12]: Done.

**Comment by shaper** [comment]
> Needs renaming

Rename is complete. Outstanding items which could be done in follow-up PRs:
- throw on `request_format: json_object`, `generateObject`, `streamObject` as it's unsupported
- rm `generate-object` and `stream-object` examples until they support `json_object`
- review/validate we do the right thing with the different function-calling modes
- review/update model feature set table to reflect manual checks to verify operation

File: content/providers/01-ai-sdk-providers/22-xai-grok.mdx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+## Model Capabilities
+
+xAI Grok offers [a variety of models with different
+capabilities](https://console.x.ai/) under "View models", including:
+
+| Model       | Image Input         | Object Generation   | Tool Usage          | Tool Streaming      |
+| ----------- | ------------------- | ------------------- | ------------------- | ------------------- |
+| `grok-beta` | <Cross size={18} /> | <Check size={18} /> | <Check size={18} /> | <Check size={18} /> |
```
Comment (lgrammel) [inline content/providers/01-ai-sdk-providers/22-xai-grok.mdx:99]: are there any limitations in their tool usage?

File: packages/xai/package.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    "./internal": {
+      "types": "./internal/dist/index.d.ts",
+      "import": "./internal/dist/index.mjs",
+      "module": "./internal/dist/index.mjs",
+      "require": "./internal/dist/index.js"
+    }
+  },
+  "dependencies": {
+    "@ai-sdk/provider": "1.0.0-canary.0",
+    "@ai-sdk/provider-utils": "2.0.0-canary.3"
```
Comment (lgrammel) [inline packages/xai/package.json:41]: This needs to be exactly the current version in the repo or the changesets updates won't work.

File: packages/xai/src/convert-to-xai-chat-messages.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                return {
+                  type: 'image_url',
+                  image_url: {
+                    url:
+                      part.image instanceof URL
+                        ? part.image.toString()
+                        : `data:${
+                            part.mimeType ?? 'image/jpeg'
+                          };base64,${convertUint8ArrayToBase64(part.image)}`,
+                  },
```
Comment (lgrammel) [inline packages/xai/src/convert-to-xai-chat-messages.ts:43]: Does grok support images?

File: packages/xai/src/xai-chat-settings.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  /**
+Whether to enable parallel function calling during tool use. Default to true.
+   */
+  parallelToolCalls?: boolean;
+
+  /**
+A unique identifier representing your end-user, which can help OpenAI to
+monitor and detect abuse. Learn more.
+*/
+  user?: string;
```
Comment (lgrammel) [inline packages/xai/src/xai-chat-settings.ts:14]: Do they support these options?

**Comment by lgrammel** [comment]
Regarding docs pages: we need to setup a redirect from the old xai docs page to the new one (in the ai studio repo). cc @jeremyphilemon for details

**Review by lgrammel — APPROVED** [review]
Have not fully tested it. Let's make sure it's fully functional before landing.

File: content/providers/01-ai-sdk-providers/22-xai-grok.mdx (hunk 5 above)
Comment (shaper) [reply content/providers/01-ai-sdk-providers/22-xai-grok.mdx:99]: Not that I know of currently. Tool streaming just means tool calls work with `streamText()`?

File: packages/xai/package.json (hunk 6 above)
Comment (shaper) [reply packages/xai/package.json:41]: I just double checked this and above and both seem to match, am I missing something?

File: packages/xai/src/convert-to-xai-chat-messages.ts (hunk 7 above)
Comment (shaper) [reply packages/xai/src/convert-to-xai-chat-messages.ts:43]: It doesn't seem to currently:
- nothing in API docs
- attempting with the `generate-text` example produces: `'{"code":"Client specified an invalid argument","error":"The model does not support content elements of type image."}'`

Plan:
- edit docs to turn the check to a cross for image support
- change conversion code to behave like other providers that don't support images (presuming we throw but will validate/emulate)
- @lgrammel - let them know we don't see support for images on the shared Slack channel (or let me know if I should do it, best to have one person talking w/ them though IMO)

File: packages/xai/src/xai-chat-settings.ts (hunk 8 above)
Comment (shaper) [reply packages/xai/src/xai-chat-settings.ts:14]: - `parallelToolCalls`: not documented in their API, attempt to prompt multiple tool calls in one response not initially successful
- `user`: yes, it's noted as an optional field [here](https://docs.x.ai/api/endpoints#chat-completions)
- `downloadImages`: only helpful if model can take image data as input, which it doesn't appear to support today

Plan:
- `parallelToolCalls`: remove
- `user`: leave as-is (and add it to their provider page, I see it's on Azure's for example)
- `downloadImages`: remove

**Comment by shaper** [comment]
> > Needs renaming
> 
> Rename is complete. Outstanding items which could be done in follow-up PRs:
> 
> * throw on `request_format: json_object`, `generateObject`, `streamObject` as it's unsupported

disabling object-generation-mode effectively did this, though we now also throw unsupported-functionality in the appropriate case.

> * rm `generate-object` and `stream-object` examples until they support `json_object`

Leaving these for now as they're useful to check/test (and presumably they'll likely add support sometime). If you'd prefer removal just let me know.

> * review/validate we do the right thing with the different function-calling modes

I did this manually and they all functioned as I expect.

> * review/update model feature set table to reflect manual checks to verify operation

Done.

File: packages/xai/src/xai-chat-language-model.ts [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+  headers: () => Record<string, string | undefined>;
+  url: (options: { modelId: string; path: string }) => string;
+  fetch?: FetchFunction;
+};
+
+export class XaiChatLanguageModel implements LanguageModelV1 {
+  readonly specificationVersion = 'v1';
+
+  readonly supportsStructuredOutputs = false;
+  readonly defaultObjectGenerationMode = undefined;
```
Comment (lgrammel) [inline packages/xai/src/xai-chat-language-model.ts:39]: they seem to support generation in `tool` mode (at least implemented below). if that works let's switch the default to `tool`

**Review by lgrammel — APPROVED** [review]

File: content/providers/01-ai-sdk-providers/22-xai-grok.mdx (hunk 5 above)
Comment (lgrammel) [reply content/providers/01-ai-sdk-providers/22-xai-grok.mdx]: this is great, it means we can set object generation to `tool` mode with forced function calling

File: packages/xai/package.json (hunk 6 above)
Comment (lgrammel) [reply packages/xai/package.json:41]: ah yes, you're right, got confused by the low canary numbers

**Pull request merged**